File: 프로그래머스/2/169198. 당구 연습/당구 연습.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

vector<int> solution(int m, int n, int startX, int startY, vector<vector<int>> balls) {
    vector<int> answer;
    
    int distanceX = 0, distanceY = 0, tmp = 0;
    for(auto ball: balls) {
        int distance = 2e9;
        distanceX = pow(startX - ball[0], 2);
        distanceY = pow(startY - ball[1], 2);
        
        if(startX != ball[0]) {
            tmp = distanceX + pow(startY + ball[1], 2); // 아래 쿠션
            distance = min(distance, tmp); 
            
            tmp = distanceX + pow((n - startY) + (n - ball[1]), 2); // 위 쿠션
            distance = min(distance, tmp);
        }
        else {
            if(startY < ball[1]) {
                tmp = distanceX + pow(startY + ball[1], 2); // 아래 쿠션
                distance = min(distance, tmp); 
            }
            else {
                tmp = distanceX + pow((n - startY) + (n - ball[1]), 2); // 위 쿠션
                distance = min(distance, tmp);
            }
        }
        
        if(startY != ball[1]) {
            tmp = distanceY + pow(startX + ball[0], 2); // 왼 쿠션
            distance = min(distance, tmp); 
            
            tmp = distanceY + pow((m - startX) + (m - ball[0]), 2); // 오른 쿠션
            distance = min(distance, tmp);
        }
        else {
            if(startX < ball[0]) {
                tmp = distanceY + pow(startX + ball[0], 2); // 왼 쿠션
                distance = min(distance, tmp);
            }
            else {
                tmp = distanceY + pow((m - startX) + (m - ball[0]), 2); // 오른 쿠션
                distance = min(distance, tmp);
            }
        }
        
        answer.push_back(distance);
    }
    
    return answer;
}

File: 프로그래머스/2/169198. 당구 연습/당구 연습.js
function solution(m, n, startX, startY, balls) {
    let answer = [];
    let distanceX, distanceY, check;
    
    for(const ball of balls) {
        let distance = Infinity;
        
        let distanceX = (startX - ball[0]) ** 2;
        let distanceY = (startY - ball[1]) ** 2;
        
        // (원쿠션 거리)^2: (두 공 사이의 거리)^2 + (두 공의 쿠션할 벽까지의 거리합)^2 
        
        if(startX !== ball[0]) {
            check = distanceX + (startY + ball[1]) ** 2; // 아래 쿠션
            distance = Math.min(distance, check);
            
            check = distanceX + ((n - startY) + (n - ball[1])) ** 2; // 위 쿠션
            distance = Math.min(distance, check);
        } else {
            // X 좌표가 같으면 쿠션하다가 부딪치기 때문에 따로 관리
            if(startY < ball[1]) {
                check = distanceX + (startY + ball[1]) ** 2; // 아래 쿠션
                distance = Math.min(distance, check);
            } else {
                check = distanceX + ((n - startY) + (n - ball[1])) ** 2; // 위 쿠션
                distance = Math.min(distance, check);
            }
        }
        
        if(startY !== ball[1]) {
            check = distanceY + (startX + ball[0]) ** 2; // 왼 쿠션
            distance = Math.min(distance, check);
            
            check = distanceY + ((m - startX) + (m - ball[0])) ** 2; // 오른 쿠션
            distance = Math.min(distance, check);
        } else {
            // Y 좌표가 같으면 쿠션하다가 부딪치기 때문에 따로 관리
            if(startX < ball[0]) {
                check = distanceY + (startX + ball[0]) ** 2; // 왼 쿠션
                distance = Math.min(distance, check);
            } else {
                check = distanceY + ((m - startX) + (m - ball[0])) ** 2; // 오른 쿠션
                distance = Math.min(distance, check);
            }
        }
        
        answer.push(distance);
    }
    
    return answer;
}